### ISSUE TITLE
boost 1.55 : Reversed (or previously applied) patch detected ? 

### ISSUE THREAD

**Issue by pramodskumbhar** [spack/spack#5325]
While installing boost 1.55 get:

```
$ spack install -v 'boost@1.55.0+python^python@2.7.13' %gcc@5.3.0 arch=linux-rhel6-x86_64
==> python is already installed in /gpfs/bbp.cscs.ch/scratch/gss/bgq/kumbhar-adm/SPACK_HOME/install_home/bbpviz/install/linux-rhel6-x86_64/gcc-5.3.0/python-2.7.13-dhkjobme
==> bzip2 is already installed in /gpfs/bbp.cscs.ch/scratch/gss/bgq/kumbhar-adm/SPACK_HOME/install_home/bbpviz/install/linux-rhel6-x86_64/gcc-5.3.0/bzip2-1.0.6-npjovari
==> zlib is already installed in /gpfs/bbp.cscs.ch/scratch/gss/bgq/kumbhar-adm/SPACK_HOME/install_home/bbpviz/install/linux-rhel6-x86_64/gcc-5.3.0/zlib-1.2.11-cabzx2gn
==> Installing boost
==> Fetching http://downloads.sourceforge.net/project/boost/boost/1.55.0/boost_1_55_0.tar.bz2
######################################################################## 100.0%
==> Staging archive: /gpfs/bbp.cscs.ch/home/kumbhar-adm/SPACK_HOME/spack/var/spack/stage/boost-1.55.0-v4npzsjwsspjcvrf4stbswabdfvq7kwd/boost_1_55_0.tar.bz2
==> Created stage in /gpfs/bbp.cscs.ch/home/kumbhar-adm/SPACK_HOME/spack/var/spack/stage/boost-1.55.0-v4npzsjwsspjcvrf4stbswabdfvq7kwd
==> Applied patch call_once_variadic.patch
Reversed (or previously applied) patch detected!  Assume -R? [n] ^C==> Patch boost_10125.patch failed.
```

EDIT: 1.57 doesn't have this issue.

**Comment by aprokop** [comment]
I added the first fix in 1128922b444d3390538b2da65937263d9c8a0b50. However, later, @hartzell added the same patch in 18c7e9662437081e4d283abf656e1604228af045 without realizing it. Essentially, we just need to leave a single patch and make certain of the exact boost and gcc versions for it.

**Comment by pramodskumbhar** [comment]
thanks for info. 

I wonder if I haven't enabled `+python` then why `python_jam_pre156.patch` getting applied: 
```
spack install -v 'boost@1.50.0' %gcc@5.3.0 arch=linux-rhel6-x86_64
....
==> Applied patch python_jam_pre156.patch
3 out of 3 hunks FAILED -- saving rejects to file boost/thread/pthread/once.hpp.rej
```
Seems like many older versions are not installable now.

**Comment by aprokop** [comment]
@pramodskumbhar Regarding python, I thinks this is a second bug, and should be changed to
```
-    patch('python_jam.patch', when='@1.56.0: ^python@3:')
-    patch('python_jam_pre156.patch', when='@:1.55.0 ^python@3:')
+    patch('python_jam.patch', when='@1.56.0:+python ^python@3:')
-+    patch('python_jam_pre156.patch', when='@:1.55.0+python ^python@3:')
```

**Comment by tgamblin** [comment]
@junghans: does @aprokop's fix address #5341?

**Comment by junghans** [comment]
@tgamblin: I think it solves 50% of the issue.

I didn't run into the problem with `python_jam*.patch`.

**Comment by alalazo** [comment]
Apparently fixed by #5934
